filament/backend/src/vulkan/VulkanFboCache.cpp

```diff
@@ -215,53 +198,54 @@ VkRenderPass VulkanFboCache::getRenderPass(RenderPassKey config) noexcept {
             continue;
         }
-        const VkImageLayout subpassLayout = colorLayouts[i].subpass;
+
         uint32_t index;
 
         if (!hasSubpasses) {
             index = subpasses[0].colorAttachmentCount++;
-            colorAttachmentRefs[0][index].layout = subpassLayout;
+            colorAttachmentRefs[0][index].layout = VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;
             colorAttachmentRefs[0][index].attachment = attachmentIndex;
         } else {
-
-            // The Driver API consolidates all color attachments from the first and second subpasses
-            // into a single list, and uses a bitmask to mark attachments that belong only to the
-            // second subpass and should be available as inputs. All color attachments in the first
-            // subpass are automatically made available to the second subpass.
-
-            // If there are subpasses, we require the input attachment to be the first attachment.
-            // Breaking this assumption would likely require enhancements to the Driver API in order
-            // to supply Vulkan with all the information needed.
-            assert_invariant(config.subpassMask == 1);
-
             if (config.subpassMask & (1 << i)) {
+                // An attachment used as both an input attachment and a color attachment must be in the LAYOUT_GENERAL.
                 index = subpasses[0].colorAttachmentCount++;
-                colorAttachmentRefs[0][index].layout = subpassLayout;
+                colorAttachmentRefs[0][index].layout = VK_IMAGE_LAYOUT_GENERAL;
                 colorAttachmentRefs[0][index].attachment = attachmentIndex;
 
                 index = subpasses[1].inputAttachmentCount++;
-                inputAttachmentRef[index].layout = subpassLayout;
+                inputAttachmentRef[index].layout = VK_IMAGE_LAYOUT_GENERAL;
                 inputAttachmentRef[index].attachment = attachmentIndex;
-            }
 
-            index = subpasses[1].colorAttachmentCount++;
-            colorAttachmentRefs[1][index].layout = subpassLayout;
-            colorAttachmentRefs[1][index].attachment = attachmentIndex;
+                index = subpasses[1].colorAttachmentCount++;
+                colorAttachmentRefs[1][index].attachment = VK_ATTACHMENT_UNUSED;
+            }
+            else {
+                index = subpasses[1].colorAttachmentCount++;
+                colorAttachmentRefs[1][index].layout = VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;
+                colorAttachmentRefs[1][index].attachment = attachmentIndex;
+            }
         }
 
         const TargetBufferFlags flag = TargetBufferFlags(int(TargetBufferFlags::COLOR0) << i);
         const bool clear = any(config.clear & flag);
-        const bool discard = any(config.discardStart & flag);
+        const bool discardStart = any(config.discardStart & flag);
+
+        // The mission here is to ensure that the initial layout matches the current layout of the image.
```

Comment:
It might be better to track layout than to try to deduce it with difficult logic.